src/MALA.jl

```diff
@@ -0,0 +1,103 @@
+using ForwardDiff: gradient!
+using DiffResults: GradientResult, value, gradient
+
+struct MALA{D} <: MHSampler
+    proposal::D
+end
+
+
+MALAMH(d) = MALA(RandomWalkProposal(d))
+
+
+struct Transition_w∇{T<:Union{Vector, Real, NamedTuple}, L<:Real, G<:Union{Vector, Real, NamedTuple}}
+    params :: T
+    lp :: L
+    ∇ :: G
+end
+
+
+# Store the new draw, its log density and its gradient
+Transition_w∇(model::DensityModel, params) = Transition_w∇(params, ∂ℓπ∂θ(model, params)...)
+
+
+function propose(
+    rng::Random.AbstractRNG,
+    spl::MALA{<:Proposal},
+    model::DensityModel,
+    params_prev::Transition_w∇
+)
+    proposal = propose(rng, spl.proposal(params_prev.∇), model, params_prev.params)
+    return Transition_w∇(model, proposal)
+end
+
+
+function q(
+    spl::MALA{<:Proposal},
+    t::Transition_w∇,
+    t_cond::Transition_w∇
+)
+    return q(spl.proposal(-t_cond.∇), t.params, t_cond.params)
+end
+
+
+"""
+    ∂ℓπ∂θ(model::DensityModel, θ::T)
+
+Efficiently returns the value and gradient of the model
+"""
+function ∂ℓπ∂θ(model::DensityModel, params)
+    res = GradientResult(params)
+    gradient!(res, model.logdensity, params)
+    return (value(res), gradient(res))
+end
+
+
+logdensity(model::DensityModel, t::Transition_w∇) = t.lp
+
+
+# Define the first step! function, which is called at the 
+# beginning of sampling. Return the initial parameter used
+# to define the sampler.
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::MALA{<:Proposal},
+    N::Integer,
+    ::Nothing;
+    init_params=nothing,
+    kwargs...
+)
+    if init_params === nothing
+        @warn "need init_params"
+    else
+        return Transition_w∇(model, init_params)
+    end
+end
+
+
+
+# Define the other step functions. Returns a Transition containing
+# either a new proposal (if accepted) or the previous proposal 
+# (if not accepted).
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::MALA{<:Proposal},
+    ::Integer,
+    params_prev;
+    kwargs...
+)
+    # Generate a new proposal.
+    params = propose(rng, spl, model, params_prev)
+
+    # Calculate the log acceptance probability.
+    logα = logdensity(model, params) - logdensity(model, params_prev) +
+        q(spl, params_prev, params) - q(spl, params, params_prev)
+
+    # Decide whether to return the previous params or the new one.
+    if -Random.randexp(rng) < logα
+        return params
+    else
+        return params_prev
+    end
+end
```

Comment:
Since this whole section is equivalent to the code [here](https://github.com/TuringLang/AdvancedMH.jl/blob/62b638ebc523b5d114fbf10476e08f0ac02cedad/src/mh-core.jl#L194), can you expand the type constraints on line 197 of `src/mh-core.jl` from

```julia
    spl::MetropolisHastings,
```

to 

```julia
    spl::Union{MetropolisHastings, MALA},
```

Reply:
Then you can remove this whole function.

Reply:
Great idea.

My apologies, but this is my first ever (!) pull request to a collaborative repo. How do I go about making edits of this scale within the pull request. Or, do I edit the fork of the repo on my end, and then resubmit a request.

My apologies for my github-ignorant question.

Reply:
You can just push them right to this branch, no need to worry about resubmitting a PR.